store/database.go

```diff
@@ -0,0 +1,78 @@
+package store
+
+import (
+	"io"
+)
+
+// VersionedReader extends the Reader interface by making all reads versioned.
+type VersionedReader interface {
+	Has(storeKey string, version uint64, key []byte) (bool, error)
+	Get(storeKey string, version uint64, key []byte) ([]byte, error)
+}
+
+// Reader wraps the Has and Get method of a backing data store.
+type Reader interface {
+	// Has retrieves if a key is present in the key-value data store.
+	//
+	// Note: <key> is safe to modify and read after calling Has.
+	Has(storeKey string, key []byte) (bool, error)
+
+	// Get retrieves the given key if it's present in the key-value data store.
+	//
+	// Note: <key> is safe to modify and read after calling Get.
+	// The returned byte slice is safe to read, but cannot be modified.
+	Get(storeKey string, key []byte) ([]byte, error)
+}
+
+// VersionedWriter extends the Writer interface by making all writes versioned.
+type VersionedWriter interface {
+	Set(storeKey string, version uint64, key, value []byte) error
+	Delete(storeKey string, version uint64, key []byte) error
+	GetLatestVersion() (uint64, error)
```

Comment:
is it a Reader API?

Reply:
See https://github.com/cosmos/cosmos-sdk/pull/17113#discussion_r1284605178. I think we should move the versioned interfaces to `storage`, i.e. SS. WDYT?

Reply:
yeah, I'd like to keep it in the base layer, since I think the SC has almost the same interface in terms of CRUD 
but VersionedWriter doesn't make sense in terms of SC

Reply:
Actually, the `Versioned` interfaces are probably going to be in `core` -- this is what @tac0turtle requested. This is so apparently other devs can write their own SS backends w/o relying on `store` package.